### DISCUSSION TITLE
Feedback Cues

### DISCUSSION THREAD

**Discussion by sofia-bg** [UNC-Project-Open-AAC/OS-DPI#203] [Ideas]
Moving this conversation out of issues. I am wondering if we could walk through the CSS to make new cues. With the circle trying to figure out what lines of code are making it a circle as I think about other shapes/fill patterns. 

Also, is there a way that we could have a shape (dot etc.) that represents where the cursor is moving? With mouse emulation on Microsoft 10 I was able to hide the cursor so does not seem like it would be a problem with seeing the system mouse and a feedback cue we made to track the movement across the screen. 

@gbishop @aetkie

**Comment by gbishop** [comment]
Yes, I'll be happy to provide a tour of the CSS. It is something of an art for sure. Lots to know. 

That circle Cue is particularly hairy. I mostly did that one to show it could be done and as I vaguely recall I copied some stuff and then monkeyed with it until I got something that looked OK. Wow conic-gradient? What is that?

I am no CSS expert but there are lots of examples online we can copy from. 

I don't follow your second paragraph. I'll need an example or a video chat to understand.

**Reply by sofia-bg** [reply]
This is an example of the three cues Dawn Sowers used in her dissertation. The "pointer" is what I guess I was trying to describe, it moves across the screen just the way the cursor does when using the mouse, as opposed to hopping from cell to cell and appearing in the middle. 


![Screenshot 2023-06-22 at 11 29 28 AM](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/92817124/83ed8bda-ee3e-4f89-84c6-529f7a827980)

**Comment by gbishop** [comment]
Isn't that what the mouse cursor does? How does this differ?

**Reply by sofia-bg** [reply]
Yes it behaves the same as the cursor, but it is a standard feature in current devices for it to be a dot. On the surface pro the only options available seem to be different iterations of a standard cursor in either black or white and slightly different sizes. I am interested in doing some work replicating and extending Dawn's work so it would be nice to be able to emulate the red dot instead of using the cursor's available on the device itself.

**Comment by gbishop** [comment]
Ok, thanks for the clarification. I believe we can easily define our own
mouse cursor. I'll look into it.

**Comment by gbishop** [comment]
Add a Customize component to your design with this content:
`* { cursor:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABhWlDQ1BJQ0MgcHJvZmlsZQAAKJF9kT1Iw1AUhU9btSIVUYuIOGSoTnZREcdahSJUCLVCqw4mL/2DJg1Jiouj4Fpw8Gex6uDirKuDqyAI/oC4ujgpukiJ9yWFFjFeeLyP8+45vHcf4K+XmWp2xABVs4xUIi5ksqtC8BVdGEI/fBiQmKnPiWISnvV1T31Ud1Ge5d33Z/UqOZMBPoE4xnTDIt4gntm0dM77xGFWlBTic+IJgy5I/Mh12eU3zgWH/TwzbKRT88RhYqHQxnIbs6KhEk8TRxRVo3x/xmWF8xZntVxlzXvyF4Zy2soy12mNIoFFLEGEABlVlFCGhSjtGikmUnQe9/CPOH6RXDK5SmDkWEAFKiTHD/4Hv2dr5qcm3aRQHOh8se2PMSC4CzRqtv19bNuNEyDwDFxpLX+lDsx+kl5raZEjoG8buLhuafIecLkDDD/pkiE5UoCWP58H3s/om7LA4C3Qs+bOrXmO0wcgTbNK3gAHh8B4gbLXPd7d3T63f3ua8/sBO15ykFMlJikAAAAJcEhZcwAALiMAAC4jAXilP3YAAAAHdElNRQfnBhYRDA5rTtc1AAAAGXRFWHRDb21tZW50AENyZWF0ZWQgd2l0aCBHSU1QV4EOFwAAAM5JREFUOMvFkzEKAjEQRZ8L4VsIlnsBL2En2u4RPIHWloKXsVcQ7CxEsPIGWy+2doaB1SaChlV3QXAgRRL+m8mfCfw7WlWHJrWBDBgADiiAjfP+9JVo0siks0m3irUyqfO2ApNGwDZkfRdHYOi8vwIkUdnLL2KAPjB7bJKniwxIa3o3rQIMGpifmtSLAa5hB7sxoGggvgF5DNg0AByc95cXQBiSdQ1xCSyqTAQYhz5/Ek+c97tPk9g2aR5NY2nS3qRhrb/wBOsFt/PHm38ed9YARR8YM5IgAAAAAElFTkSuQmCC), auto;
}
`

I'm defining a custom cursor using a 16x16 image. I drew it in the image editor I have installed, and saved it as a .png file. Then I used an [online converter](https://www.site24x7.com/tools/image-to-datauri.html) to get the [data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) for it. Finally I pasted that into the Customize block wrapped like this:

`* { cursor: url(data:...), auto; }`

The asterisk means match every element then the next part is the style I want to apply to every element, namely the custom cursor. You could have different cursors over different components if you wanted.

I believe SVG images might work as well as PNG. You could even draw something in excalidraw, export and convert it.

**Comment by gbishop** [comment]
Here is a version in SVG which would be nicer to hack on directly.
`* { cursor:url('data:image/svg+xml;utf8,<svg height="16" width="16" viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"> <circle cx="50" cy="50" r="40" fill="red"/> </svg>'), auto;}`

The height and width determine the size on the screen. The viewbox determines the virtual coordinate system of the image. You can adjust the color simply by changing the fill.

**Reply by sofia-bg** [reply]
awesome - thank you!

**Reply by gbishop** [reply]
You can make it transparent also with `fill="rgb(100% 0% 0% / 50%)"` for example.

**Comment by sofia-bg** [comment]
@gbishop 

Finally getting back to this, all really helpful! Have the red cursor working on my end and able to change size/color etc!

Working on having magnify work as the cursor moves. I have been able to change the cursor so that it looks like a magnifying glass, but all of the CSS code I am finding to magnify the image on the screen isn't working. Any ideas? 

What I am trying to emulate is having the target cell someone is looking at, and only that cell, grow by 50% and have this happen as the cursor moves across the screen. Will keep working on for a bit longer so will post if I get something working.

**Reply by sofia-bg** [reply]
is it because the cells aren't considered images and the zoom css codes only work on images rather than the screen generally?

This plug in works but optimally would be just the cell and a square shape. 

https://chrome.google.com/webstore/detail/magnifying-glass-hover-zo/gfcbebjdigncefnokoncehghgfndjcgn/related

**Comment by gbishop** [comment]
I've done the zoom effect. I'll have to remember how. I *think* I used
https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/scale

I have a meeting now. Will look later.

**Reply by gbishop** [reply]
Check out this video. The CSS is transform: zoom(2);
[Screencast from 2023-11-02 14-20-58.webm](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/198012/84fae4ef-4702-400e-964e-6c9011d3d529)

**Reply by sofia-bg** [reply]
oh awesome - that is exactly what i was looking for! thank you!
